File: Baekjoon/4386_별자리 만들기/B_4386.py
import math
import sys


def union(x, y):
    px = find_set(x)
    py = find_set(y)

    p[py] = px


def find_set(x):
    if p[x] == x:
        return x
    p[x] = find_set(p[x])
    return p[x]


def make_set(x):
    p[x] = x


def get_distance(a, b):
    y1, x1 = a
    y2, x2 = b

    distance = math.sqrt((y2 - y1) ** 2 + (x2 - x1) ** 2)
    return distance


n = int(sys.stdin.readline().rstrip())
arr = [list(map(float, sys.stdin.readline().split())) for _ in range(n)]

new_arr = []
for i in range(n - 1):
    for j in range(i + 1, n):
        dist = get_distance(arr[i], arr[j])
        new_arr.append([dist, i, j])

p = [0] * (n + 1)
for i in range(n + 1):
    make_set(i)
new_arr.sort()

sums = 0
cnt = 0
for i in range(len(new_arr)):
    d, a, b = new_arr[i]
    if find_set(a) == find_set(b):
        continue

    sums += d
    union(a, b)

    cnt += 1
    if cnt == len(new_arr) - 1:
        break

sys.stdout.write(str(sums))

# 정수가 아닌 소수 좌표들을 어떻게 처리할 것인지
# 소수를 어떻게 인덱스화 해야 할지
# => 입력받은 순서를 인덱스화
# 거리, 출발점, 도착점

File: Baekjoon/4386_별자리 만들기/B_4386.js
// const fs = require('fs')
// const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n')

const unionSet = function(firstItem, secondItem) {
    const px = findSet(firstItem)
    const py = findSet(secondItem)

    p[py] = px
}

const findSet = function(x) {
    if (p[x] === x) {
        return p[x]
    }

    p[x] = findSet(p[x])
    return p[x]
}

const getDistance = function(first, second) {
    const [firstY, firstX] = first
    const [secondY, secondX] = second

    const sqrtValue = Math.sqrt(Math.pow(Math.abs(firstY - secondY), 2) + Math.pow(Math.abs(firstX - secondX), 2))
    return sqrtValue
}

const input = [
'3',
'1.0 1.0',
'2.0 2.0',
'2.0 4.0'
]

const n = parseInt(input[0])
let coordinates = []

for (let i = 1; i <= n; i++) {
    const [y, x] = input[i].split(' ').map(v => parseInt(v))
    coordinates.push([y, x])
}

let distances = []

for (let i = 0; i < n - 1; i++) {
    for (let j = i + 1; j < n; j++) {
        const sqrtVal = getDistance(coordinates[i], coordinates[j])
        const distance = [coordinates[i], coordinates[j], sqrtVal]
        distances.push(distance)
    }
}

distances.sort((first, second) => {
    return first[2] - second[2]
})


let p = new Array(distances.length).fill(0)
for (let i = 1; i <= distances.length; i++) {
    p[i] = i
}

let cnt = 0
let sums = 0

for (let i = 0; i < distances.length; i++) {
    const [start, end, value] = distances[i]

    if (findSet(start) === findSet(end)) {
        continue
    }

    unionSet(start, end)
    sums += value

    cnt += 1
    if (cnt === distances.length - 1) {
        break
    }
}

console.log(sums)
